File: bench/fannkuch/fannkuch.lua
function fannkuch(n)
   local perm, perm1, count, maxPerm = {}, {}, {}, {}
   local maxFlipsCount, m, r, check = 0, n-1, n, 0

   for i=0,m do perm1[i] = i end   
   while true do
      if check < 30 then
         for i=0,m do io.write(perm1[i]+1) end; io.write("\n")
         check = check + 1;
      end

      while r ~= 1 do count[r-1] = r; r = r-1 end           
      if not (perm1[0] == 0 or perm1[m] == m) then
         for i=0,m do perm[i] = perm1[i] end
         local flipsCount, k = 0
         while true do        
            k = perm[0]                       
            if k == 0 then break end                                                               
            for i=0,(k+1)/2 - 1 do perm[i],perm[k-i] = perm[k-i],perm[i] end
            flipsCount = flipsCount + 1
         end

         if flipsCount > maxFlipsCount then
            maxFlipsCount = flipsCount
            for i=0,m do maxPerm[i] = perm1[i] end
         end
      end

      while true do
         if r == n then return maxFlipsCount end
         local perm0, i, j = perm1[0], 0       
         while i < r do
            j = i + 1
            perm1[i] = perm1[j]
            i = j
         end
         perm1[r] = perm0  
               
         count[r] = count[r] - 1
         if count[r] > 0 then break end
         r = r + 1
      end
   end
end
 
n = tonumber(arg and arg[1]) or 1
io.write("Pfannkuchen(", n, ") = ", fannkuch(n), "\n")

File: bench/fannkuch/fannkuch.go
/*
 * The Computer Language Benchmarks Game
 * http://shootout.alioth.debian.org/
 *
 * contributed by The Go Authors.
 * Based on fannkuch.c by Heiner Marxen
 * flag.Arg hack by Isaac Gouy
 */

package main

import (
   "flag"
   "fmt"
   "strconv"
)

var n = 0

func fannkuch(n int) int {
   if n < 1 {
      return 0
   }

   n1 := n - 1
   perm := make([]int, n)
   perm1 := make([]int, n)
   count := make([]int, n)

   for i := 0; i < n; i++ {
      perm1[i] = i   // initial (trivial) permutation
   }

   r := n
   didpr := 0
   flipsMax := 0
   for {
      if didpr < 30  {
         for i := 0; i < n; i++ {
            fmt.Printf("%d", 1+perm1[i])
         }
         fmt.Printf("\n")
         didpr++
      }
      for ; r != 1; r-- {
         count[r-1] = r
      }

      if perm1[0] != 0 && perm1[n1] != n1 {
         flips := 0
         for i := 1; i < n; i++ {   // perm = perm1
            perm[i] = perm1[i]
         }
         k := perm1[0]      // cache perm[0] in k
         for {         // k!=0 ==> k>0
            for i, j := 1, k-1; i < j; i, j = i+1, j-1 {
               perm[i], perm[j] = perm[j], perm[i]
            }
            flips++
            // Now exchange k (caching perm[0]) and perm[k]... with care!
            j := perm[k]; perm[k] = k; k = j;
            if k == 0 {
               break
            }
         }
         if flipsMax < flips {
            flipsMax = flips
         }
      }

      for ; r < n; r++ {
         // rotate down perm[0..r] by one
         perm0 := perm1[0]
         for i := 0; i < r; i++ {
            perm1[i] = perm1[i+1]
         }
         perm1[r] = perm0
         count[r]--
         if count[r] > 0 {
            break
         }
      }
      if r == n {
         return flipsMax
      }
   }
   return 0
}

func main() {
   flag.Parse()
   if flag.NArg() > 0 { n,_ = strconv.Atoi( flag.Arg(0) ) }
   fmt.Printf("Pfannkuchen(%d) = %d\n", n, fannkuch(n))
}
